piet-direct2d/src/dwrite.rs

```diff
@@ -178,81 +268,113 @@ impl TextFormatBuilder {
     }
 }
 
-impl TextLayoutBuilder {
-    pub fn new(factory: DwriteFactory) -> TextLayoutBuilder {
-        TextLayoutBuilder {
-            factory,
-            format: None,
-            text: None,
-            width: None,
-            height: None,
-            alignment: TextAlignment::default(),
-        }
-    }
-
-    pub fn format(mut self, format: &TextFormat) -> TextLayoutBuilder {
-        // The fact we clone here is annoying, but it gets us out of
-        // otherwise annoying lifetime issues.
-        self.format = Some(format.clone());
-        self
-    }
-
-    pub fn text(mut self, text: &str) -> TextLayoutBuilder {
-        self.text = Some(text.to_wide());
-        self
+impl AsRef<str> for FamilyName {
+    fn as_ref(&self) -> &str {
+        &self.name
     }
+}
 
-    pub fn width(mut self, width: f32) -> TextLayoutBuilder {
-        self.width = Some(width);
-        self
+impl AsRef<[u16]> for FamilyName {
+    fn as_ref(&self) -> &[u16] {
+        &self.wide_name
     }
+}
 
-    pub fn height(mut self, height: f32) -> TextLayoutBuilder {
-        self.height = Some(height);
-        self
-    }
+#[allow(overflowing_literals)]
+#[allow(clippy::unreadable_literal)]
+const E_NOT_SUFFICIENT_BUFFER: HRESULT = 0x8007007A;
 
-    pub fn alignment(mut self, alignment: TextAlignment) -> TextLayoutBuilder {
-        self.alignment = alignment;
-        self
+fn make_text_range(start: usize, len: usize) -> DWRITE_TEXT_RANGE {
+    DWRITE_TEXT_RANGE {
+        startPosition: start as u32,
+        length: len as u32,
     }
+}
 
-    pub fn build(self) -> Result<TextLayout, Error> {
-        let format = self.format.expect("`format` must be specified");
-        let text = self.text.expect("`text` must be specified");
+impl TextLayout {
+    pub(crate) fn new(
+        dwrite: &DwriteFactory,
+        format: TextFormat,
+        width: f32,
+        text: &[u16],
+    ) -> Result<Self, Error> {
+        const QUITE_TALL_HEIGHT: f32 = 1e6;
         let len = text.len();
         assert!(len <= 0xffff_ffff);
```

Comment:
I probably wrote that assert, but more idiomatic style would be `try_into` instead of `as`.

Reply:
my inclination is to truncate instead of asserting, but agree it's a mostly theoretical concern